### PR TITLE
fix(uninstall): keep side-by-side installs that share a bundle id

### LINKED ISSUE #1186: [BUG] Xcode-beta.app never shows up in mo uninstall when Xcode.app is also installed
## Describe the bug

With both `/Applications/Xcode.app` and `/Applications/Xcode-beta.app` installed, `mo uninstall` and `mo uninstall --list` only show Xcode. The beta never appears, so there is no way to uninstall it through Mole.

The cause is the backup-volume dedupe added for #975. Apple ships Xcode and Xcode-beta with the same `CFBundleIdentifier` (`com.apple.dt.Xcode`), and `_scan_dedupe_bundle_ids` in `bin/uninstall.sh` keys rows on bundle ID alone. The beta gets collapsed into the stable install as if it were a clone on a backup volume.

```
$ plutil -extract CFBundleIdentifier raw /Applications/Xcode.app/Contents/Info.plist
com.apple.dt.Xcode
$ plutil -extract CFBundleIdentifier raw /Applications/Xcode-beta.app/Contents/Info.plist
com.apple.dt.Xcode
```

This bites hardest on macOS betas: on macOS 27 beta the stable Xcode refuses to launch, so the machine ends up with two 10GB+ installs and Mole can only see one of them.

## Steps to reproduce

1. Install Xcode and Xcode-beta side by side in `/Applications`
2. Run `mo uninstall --list | grep -i xcode`
3. Only `Xcode.app` is listed

## Expected behavior

Both installs appear in the list and each can be uninstalled by name. Mirrored clones on backup volumes (the #975 case) should still collapse.

## Environment

```text
Mole 1.44.1 · macOS 27 beta (arm64)
```

## Additional context

There is a second problem hiding behind the first one. Once both installs are visible, uninstalling only the beta would collect leftovers by bundle ID (`~/Library/Caches/com.apple.dt.Xcode`, preferences, containers, launch services), and those files still belong to the surviving stable install. A fix needs to cover both: keep the two rows apart in the scan, and skip bundle-ID-keyed leftovers while a sibling install with the same bundle ID is still on disk.

I have a fix with regression tests, PR coming right after this.

### PR DESCRIPTION
**Pull request by hhh2210**
Fixes #1186

## Summary

Xcode.app and Xcode-beta.app both ship `com.apple.dt.Xcode` as their bundle ID. The backup-volume dedupe from #975 keys scan rows on bundle ID alone, so the beta was collapsed into the stable install and never showed up in `mo uninstall`.

- `_scan_dedupe_bundle_ids` now keys on bundle ID plus the `.app` basename. True clones on mirrored volumes (the #975 case) still collapse with the live location winning, while distinct installs stay visible.
- New guard in `lib/uninstall/batch.sh`: when an unselected install with the same bundle ID is still on disk, the selected app's bundle ID is demoted to `unknown` before leftover discovery. Without this, uninstalling only the beta would trash `~/Library/Caches/com.apple.dt.Xcode`, preferences, containers, and launch services that the surviving stable install still uses. Leftovers fall back to name and path matches, which are unique per install.
- Two small fixes that surfaced while testing on this machine: `${path/$HOME/~}` in the uninstall preview became a no-op under bash 5.3 (a literal `~` in the patsub replacement is now tilde-expanded, so the replacement equals the original), routed through a variable instead; and two paginated menu tests needed `< /dev/null` because the menu's `read -n1` was eating heredoc bytes.

Dry run on the affected machine, with both Xcodes installed:

```
$ mo uninstall --list | grep -i xcode
Xcode-beta   com.apple.dt.Xcode   Xcode-beta   9.12GB
Xcode        com.apple.dt.Xcode   Xcode        12.48GB

$ mo uninstall Xcode-beta --dry-run --debug
[DEBUG] Bundle id com.apple.dt.Xcode shared with a surviving install; restricting Xcode-beta leftovers to name/path matches
  ✓ /Applications/Xcode-beta.app , 9.12GB
  ✓ ~/Library/Developer/Xcode/DerivedData
  ✓ ~/Library/Developer/CoreSimulator/Caches
```

## Safety Review

- Affects uninstall scan and leftover collection.
- The dedupe change only widens what the scan keeps; it removes nothing new. Renamed clones in backup roots (`Xcode copy.app`) no longer collapse and will show as extra rows, which seems acceptable for a scan surface.
- The sibling guard narrows deletion: bundle-ID-keyed paths are skipped whenever another install with the same bundle ID survives outside the selection. Selecting both installs restores full bundle-ID cleanup. `should_protect_path` coverage is unchanged.
- The demoted bundle ID takes the existing `unknown` code paths (`stop_launch_services`, `remove_login_item`, defaults deletion, ByHost scan all already no-op on `unknown`).

## Tests

- `MOLE_TEST_NO_AUTH=1 bats tests/uninstall.bats` (66 pass, includes 2 new tests: sibling detection unit test, full batch flow asserting shared bundle-ID files survive)
- `MOLE_TEST_NO_AUTH=1 bats tests/uninstall_scan_bash32.bats` (9 pass, includes new Xcode/Xcode-beta dedupe test; verified on /bin/bash 3.2)
- `MOLE_TEST_NO_AUTH=1 bats tests/uninstall_safety.bats tests/uninstall_remove_file_list.bats tests/uninstall_naming_variants.bats` (36 pass)
- `./scripts/check.sh` (shellcheck, syntax, format all pass)
- Manual: `mo uninstall Xcode-beta --dry-run --debug` on macOS 27 beta with both installs present, output above

## Safety-related changes

- Leftover collection now skips bundle-ID-derived paths when a same-bundle-ID install survives the uninstall. This is a deliberate narrowing to avoid deleting live data of the surviving install.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by hhh2210** [comment]
this bug is tricky and thanks fable5 for not falling back to opus)
image

File: lib/uninstall/batch.sh [hunk 1]
```diff
@@ -444,6 +447,43 @@ remove_file_list() {
...
+
+    [[ -z "$bundle_id" || "$bundle_id" == "unknown" ]] && return 1
+
+    local row other_path other_bundle
+    # shellcheck disable=SC2154 # apps_data is provided by bin/uninstall.sh via dynamic scope.
+    for row in "${apps_data[@]+"${apps_data[@]}"}"; do
+        IFS='|' read -r _ other_path _ other_bundle _ _ _ <<< "$row"
+        [[ "$other_bundle" == "$bundle_id" ]] || continue
+        [[ "$other_path" == "$app_path" ]] && continue
+        [[ -d "$other_path" ]] || continue
```
[the remark this answers is not in the document]
Comment (hhh2210) [reply lib/uninstall/batch.sh:477]: Good catch on the surface, but the main scenario here can't actually happen, and the residual one is intentional.

An exact mirror clone on a backup volume never reaches this check. The scan dedupe keys on bundle id plus `.app` basename, so `/Volumes/CCC/Applications/Xcode.app` collapses into the `/Applications/Xcode.app` row (live path wins via `path_rank`) and `apps_data` only carries one row. The guard can only see a `/Volumes` sibling when the scan already treats it as a distinct install, e.g. the live copy was deleted and only the backup remains, or the clone has a different basename.

For that residual case I kept `/Volumes` siblings counting on purpose. People do run apps directly from external volumes, and those installs use the same `$HOME` bundle-id data this guard protects. The two failure modes are not symmetric: counting a stale backup clone means a few leftover files stay behind (and `mo clean` leftover scanning can pick them up later), while ignoring `/Volumes` siblings would delete preferences and containers a real external install still uses. Mole's safety contract prefers under-deletion here.

Documented this trade-off in a comment on the helper in d44f9c4.

**Comment by hhh2210** [comment]
@tw93 Ready for review.

Went through the Copilot review: its one inline comment worried that backup clones under `/Volumes` could trip the new sibling guard. Exact mirror clones can't reach the guard (the scan dedupe already collapses them into the live row), and for the remaining corner case I kept the conservative behavior on purpose, with the reasoning documented in the code and in the reply on that thread. No logic changes needed, d44f9c4 only adds the explanatory comment.

Full test run is green: uninstall.bats 66/66, uninstall_scan_bash32.bats 9/9 (verified on /bin/bash 3.2), plus the safety suites. Dry-run output from the affected machine is in the PR description.

**Review by tw93 — CHANGES_REQUESTED** [review]
Thanks for the careful work here, this is a real bug and the overall direction is right. I did a line-by-line pass plus an adversarial second review since this touches the uninstall deletion path. The dedupe key change, the sibling guard itself, the `unknown` demotion, and the bash 3.2 hygiene all check out: I verified that every bundle-id-keyed sink (`find_app_files`, `stop_launch_services`, `defaults delete`, ByHost scan, system extensions) is reverse-DNS gated or no-ops on `unknown`, that the cache merge keys on `app_path` so both rows survive, and that mirrored-clone collapse from #975 still works. The `< /dev/null` test fixes and the tilde patsub fix are both correct.

Two gaps need to be closed before merge, both in the "sibling survives" scenario the PR is about:

**1. The Homebrew cask path bypasses the sibling guard.** `_batch_execute_removals` still runs `brew_uninstall_cask`, which does `brew uninstall --cask --zap` (`lib/uninstall/brew.sh`). Cask detection is path-based and never sees the demoted bundle id, and a zap stanza deletes exactly the bundle-id-keyed prefs/caches/Application Support the guard promises to preserve. iTerm2 is a live example: `iterm2` and `iterm2-beta` both use `com.googlecode.iterm2`, and the beta's zap trashes the stable install's preferences. When `uninstall_bundle_id_has_surviving_sibling` fires, please drop `--zap` for that cask (plain `brew uninstall --cask`) or fall back to manual bundle removal, with a mocked-brew regression test.

**2. The name/path fallback is not "unique per install" when display names collide.** `uninstall_resolve_display_name` (`bin/uninstall.sh`) falls back to CFBundleDisplayName/CFBundleName whenever `mdls` returns empty or `(null)`, which is common on unindexed volumes and beta OS builds. For Xcode-beta that yields display name "Xcode", identical to the survivor: the versioned-name guard only preserves suffixes containing a digit, and `-beta` has none. With the bundle id demoted, `find_app_files` then queues `~/Library/{Application Support,Caches,Logs,Preferences}/Xcode` and the version-suffix stripper maps "Foo Beta" back to the survivor's "Foo" dirs; `remove_login_item` also deletes login items by that shared name with no path check. That makes the new fallback branch broader than the primary branch, the exact regression shape that got #874/#875 reverted. When the guard fires, please also suppress name-derived patterns that collide with the survivor's resolved display name (restricting discovery to the `.app` basename, e.g. `Xcode-beta`, is enough), and skip `remove_login_item` on a colliding name.

**3. Test coverage.** Both new bats tests use non-colliding names ("Shared" vs "Shared-beta") and a non-cask app, so neither scenario above is covered. Please add: a batch test where both installs resolve to the same display name asserting name-keyed leftovers survive, and a mocked-brew cask sibling test asserting `--zap` is not used.

Happy to re-review quickly once these are in, the core of the change is solid.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by hhh2210** [comment]
@tw93 All three points were valid, thanks for the adversarial pass. Fixed in 77c7800.

**1. Cask zap bypass.** `brew_uninstall_cask` now takes a zap mode argument, and `_batch_execute_removals` passes `nozap` whenever the sibling guard fired for that app, so a guarded cask gets a plain `brew uninstall --cask`. The preview warning about `--zap` and the failure-path suggestion text also switch wording so we never tell the user to run the zap we just refused. Regression test: mocked-brew cask with a surviving same-bundle sibling, asserting `uninstall --cask <token>` is called and `--zap` never appears in the call log.

**2. Name collisions.** When the guard fires, leftover discovery now keys on the `.app` basename instead of the resolved display name, which stays correct in the mdls-returns-(null) case you described (Xcode-beta resolves to "Xcode" but the basename is still `Xcode-beta`). On top of that:

- If even the basename, or its version-suffix-stripped base, matches a survivor's display name, basename, or their stripped forms (case-insensitive), name discovery is suppressed entirely and only the `.app` bundle is removed. The stripped-form comparison covers the "Foo Beta" style survivor too, since `find_app_files` derives extra patterns from the stripped base.
- `remove_login_item` is skipped when the display name collides with a survivor's, since login items are matched by name only.
- One more leak your review made me re-check: the toolchain heuristics in `find_app_files` match by regex substring, so `Xcode-beta` would still sweep `DerivedData` and `* DeviceSupport` that the surviving Xcode.app uses. The guard now disables that heuristic section through `MOLE_UNINSTALL_SIBLING_SURVIVES`, gated so the normal uninstall path is byte-for-byte unchanged.

**3. Tests.** Added the two you asked for, plus one:

- `tests/uninstall.bats`: colliding display names (both rows resolve to "SharedName") asserting the survivor's name-keyed Application Support/Caches/Preferences survive and login-item removal is skipped; a same-basename variant asserting full suppression; and a toolchain case asserting `DerivedData` survives when a same-bundle sibling remains.
- `tests/brew_uninstall.bats`: the mocked-brew sibling test asserting no `--zap`.

Full uninstall/brew/app-protection suites (124 + 11 tests) pass locally, `check.sh --format` clean.

**Review by tw93 — APPROVED** [review]
Re-reviewed 77c7800 line by line plus an adversarial pass. All three earlier points are properly closed: the zap mode is derived per app from its own record so nothing leaks between batch iterations, discovery keys on the .app basename with collision suppression, and the toolchain heuristics are gated. The regression tests cover the right scenarios and assert real filesystem outcomes.

The adversarial pass surfaced one more family of paths that could still reach a surviving install through shared identifiers: the collision check was exact-match while downstream matchers are substring-based (uninstalling `Foo.app` while `Foo-beta.app` survives slipped through via the `*<name>*.plist` LaunchAgents glob), diagnostic-report discovery keys on the shared CFBundleExecutable, and process termination plus login-item-helper bootout target shared identifiers. Rather than another round trip I pushed ec1cd647 to this branch closing them, with regression tests including an inverse-direction case and a non-vacuity control; details are in the commit message.

Thanks for the careful work across all three rounds, this was a tricky one to get right.
